### ISSUE TITLE
Bash Completion

### ISSUE THREAD

**Issue by Lorentz83** [yadayada/acd_cli#127]
Hi, 
thank you for the great tool. Since I've used it so much in the last days, I've written a bash completion script for it. 
It is not completed yet, but works for the most used commands. 
If you are interested to integrate it, this is the code
https://gist.github.com/Lorentz83/cad24ca44b53e4a33626

**Comment by cpare** [comment]
@Lorentz83 - Could you give us a primer for this?  What does it do, how to use it, what's the value add?

**Comment by Lorentz83** [comment]
Of course.
It integrates better the program with bash (the default shell in most linux distributions). 
If you source this script using the command 
```
source <filename>
```
bash will "know" the semantic of the tool and will choose how auto complete the commandline whenever you press tab. 
For example, if you start typing `acd_cli do` and press tab, the command will be completed in `acd_cli download`, because "download" is the only allowed keyword in that position that begins with "do". 
Similarly, if you press tab again you'll see `acd_cli download /`, because the second parameter is a remote name. At this point you do not have a single choice anymore (unless you have only one file in the root directory of amazon cloud). Therefore if you tab twice bash will show you your remote files. Something similar to 
```
$ acd_cli download /
Documents/                Pictures/                Videos/
```

It is incomplete because, for example, the rule for `acd_cli restore` is still missing, therefore it will try to complete the second parameter with a local file (the default bash behavior) instead of completing it with a trashed file on the cloud. But for the average usage is already fairly good. 

The value added is that you can write command faster, because you can rely on a smarter auto completion. It is very useful especially when you must write remote pathnames, because you can avoid typos or cut and pastes. 

The easy way to install it is to copy the file in `/etc/bash_completion.d`
Alternately you can add the line `source  acd_cli_bash_completion` in `~/.bashrc`
The first is a system wide change, the second works only in the interactive shells of the selected user.

I hope I made myself clear

**Comment by RobinHoutevelts** [comment]
Ooh, this is mighty helpful for noobs!

**Comment by yadayada** [comment]
Thanks for the script. I will add an ls option that only outputs trashed files if it helps (and an ls option that only outputs file names). There will be the problem that trash files don't necessarily have a unique name, though.

Has anyone got this working in zsh?

**Comment by Lorentz83** [comment]
I used a lot of bash syntax, I think it will require some changes to make it work in zsh too. But I've never used that shell, so cannot be sure.

**Comment by yadayada** [comment]
It's working in my bash shells and zsh is (said to be) compatible with bash completion scripts; that's what made me wonder. Turns out I had an alias set for acdcli in my .zshrc.local, but apparently it's not possible to complete after an alias (.|?) 
After unsetting the alias, it's running and I get the first completion level, the list of actions.

```
acdcli add-child _acd_cli_in_set:shift:2: shift count must be <= $#
_acd_cli_in_set:shift:2: shift count must be <= $#
_acd_cli_in_set:shift:2: shift count must be <= $#
_acd_cli_in_set:shift:2: shift count must be <= $#
_acd_cli_in_set:shift:2: shift count must be <= $#
_acd_cli_in_set:shift:2: shift count must be <= $#
_acd_cli_in_set:shift:2: shift count must be <= $#
_acd_cli:116: command not found: compopt
```

I think ``compopt`` provides a subset of the ``complete`` builtin, so that could be a simple replacement.

The other thing that's missing from zsh is mapfile.

**Comment by Lorentz83** [comment]
compopt should be simple to replace, because it is used only to instruct bash to perform the default completions with the local files. 
The real problem is mapfile, it is the only solution that I've found to correctly handle remote filenames with spaces. 
I'll investigate better as soon as I have some time.

**Comment by yadayada** [comment]
If you tell me which output format would be optimal for the file listings, I might add a special formatter for that purpose.

**Comment by Lorentz83** [comment]
That would be great. I'm currently studying zsh. It seems a really good shell, I'm considering to make it my default shell. 
The problem is that, as far as I understood until now, even if it has a compatibility mode for bash autocompletion, it is better to migrate the scripts to use its advanced features. 
My idea is to start understanding how the compatibility mode works and fix only the minimum necessary to make the completion work. Later I'll try to understand how to use the native zsh autocompletaion.
I'd like to maintain everything only in one script, but probably it wouldn't be possible because the requirements seem very different. 

The "mapfile" related problem, is that with that command is possible to read every line into a different array entry. When the standard bultins parse the acd_cli output, they breaks the filenames which contain spaces. It may be helpful if you introduce a flag to print the names quoting the spaces. But let me investigate more before introducing this feature.

**Comment by AlexCzar** [comment]
+1 to zsh. zsh + oh-my-zsh customization suite is arguably the best shell in existence.

**Comment by zmwangx** [comment]
@yadayada 
> Turns out I had an alias set for acdcli in my .zshrc.local, but apparently it's not possible to complete after an alias (.|?) 

That's not true though, Zsh should be able to complete after an alias as if it's expanded. I don't know if that's a problem of the bash compatibility layer though. Maybe you could give my newly submitted native zsh completion a spin, in case you're interested.
